frappe/workflow/doctype/workflow_action/workflow_action.py

```diff
@@ -3,8 +3,164 @@
+		'workflow_state': doc_current_state
+	})
+
+	if workflow_action_already_created: return
+
+	immediate_next_possible_transitions = [d for d in workflow.transitions if d.state == doc_current_state]
+
+	current_user = frappe.session.user
+
+	# update status of completed workflow action of user
+	frappe.db.sql('''update `tabWorkflow Action` set status='Completed', completed_by=%s
+		where reference_doctype=%s and reference_name=%s and user=%s''',
+		(current_user, reference_doctype, reference_name, current_user))
+
+	# delete irrelevant workflow actions
+	frappe.db.sql('''delete from `tabWorkflow Action`
+		where reference_doctype=%s and reference_name=%s and user!=%s''',
+		(reference_doctype, reference_name, current_user))
+
+	clear_doctype_notifications('Workflow Action')
+
+	from frappe.permissions import has_permission
+
+	email_map = {}
+
+	for transition in immediate_next_possible_transitions:
+		users = get_users_with_role(transition.allowed)
+		for user in users:
+			if (has_approval_access(doc, workflow, user) and
+				has_permission(doctype=doc, ptype='read', user=user)): pass
+			else: continue
+
+			if not email_map.get(user):
+				email_map[user] = {
+					'possible_actions': [],
+					'email': frappe.db.get_value('User', user, 'email'),
+				}
+
+			email_map[user].get('possible_actions').append({
+				'action_name': transition.action,
+				'action_link': get_workflow_action_url(
+					transition.action,
+					reference_doctype,
+					reference_name,
+					doc_current_state)
+			})
+
+	for user in email_map.keys():
+		newdoc = frappe.get_doc({
+				'doctype': 'Workflow Action',
+				'reference_doctype': reference_doctype,
+				'reference_name': reference_name,
+				'workflow_state': doc_current_state,
+				'status': 'Open',
+				'user': user
+			})
+		newdoc.insert(ignore_permissions=True)
+
+	frappe.db.commit()
+	if email_map:
+		send_workflow_action_email(email_map, reference_doctype, reference_name)
+
+
+def send_workflow_action_email(email_map, doctype, docname):
+	email_message = _('{0}:{1}'.format(doctype, docname))
+	common_args = {
+		'template': 'workflow_action',
+		'attachments': [frappe.attach_print(doctype, docname , file_name=docname)],
+		'header': [_("Workflow Action"), "orange"],
+		'subject': _('Workflow Action')
+	}
+
+	for d in email_map.values():
+		email_args = {
+			'recipients': [d.get('email')],
+			'args': {
+				'actions': d.get('possible_actions'),
+				'message': email_message
+			},
+		}
+		email_args.update(common_args)
+		enqueue(method=frappe.sendmail, queue='short', **email_args)
+
+
+@frappe.whitelist()
+def apply_action(action, doctype, docname, current_state):
+	verify_request()
+	doc = frappe.get_doc(doctype, docname)
+	workflow = get_workflow_name(doctype)
+	doc_workflow_state = doc.get(
+		frappe.get_value('Workflow', workflow, 'workflow_state_field')
+	)
+
+	if doc_workflow_state == current_state:
+		try:
+			apply_workflow(doc, action)
+			frappe.respond_as_web_page(_("Success"),
```

Comment:
More descriptive.

`Leave Application is set to "Approved"` : `{doctype} is set to "{state}"`